sed/loader/utils.py

```diff
@@ -47,3 +47,53 @@ def gather_files(
         raise
 
     return files
+
+
+def gather_flash_files(
+    run_number: int,
+    daq: str,
+    raw_data_dir: str,
+    extension: str = "h5",
+) -> List[Path]:
+    """Returns all filenames of given run located in directory
+    for the given daq."""
+    stream_name_prefixes = {
+        "pbd": "GMD_DATA_gmd_data",
+        "pbd2": "FL2PhotDiag_pbd2_gmd_data",
+        "fl1user1": "FLASH1_USER1_stream_2",
+        "fl1user2": "FLASH1_USER2_stream_2",
+        "fl1user3": "FLASH1_USER3_stream_2",
+        "fl2user1": "FLASH2_USER1_stream_2",
+        "fl2user2": "FLASH2_USER2_stream_2",
+    }
+
+    files = sorted(
+        Path(raw_data_dir).glob(
+            f"{stream_name_prefixes[daq]}_run{run_number}_*." + extension,
+        ),
+        key=lambda filename: str(filename).rsplit("_", maxsplit=1)[-1],
+    )
+    if not files:
+        raise NoFilesFoundError(
+            f"No files found for run {run_number} in directory {raw_data_dir}",
+        )
+    return files
+
+
+def parse_h5_keys(h5_file, prefix=""):
+    """Helper method which parses the channels present in the h5 file"""
+    file_channel_list = []
+
+    for key in h5_file.keys():
+        try:
+            [
+                file_channel_list.append(s)
+                for s in parse_h5_keys(h5_file[key], prefix=prefix + "/" + key)
+            ]
```

Comment:
Here, the recusion stops automatically when there are no nested keys. That's because the recursive calls are made on the values associated with the keys, and if a value does not have any nested keys, the recursion stops for that particular level. We can of course set tests to see if this behavior breaks.

Reply:
At least with my h5 version, the call to h5file. Keys() throws an error, if there are no child keys. which is re raised in the except case and the execution fails.